plugin/pkg/admission/resourcequota/admission.go

```diff
@@ -190,52 +190,66 @@ func IncrementUsage(a admission.Attributes, status *api.ResourceQuotaStatus, cli
+			if !set[resourceName] {
+				continue
 			}
-			oldCPU := resourcequotacontroller.PodCPU(oldPod)
-			oldMemory := resourcequotacontroller.PodMemory(oldPod)
-			deltaCPU = resource.NewMilliQuantity(deltaCPU.MilliValue()-oldCPU.MilliValue(), resource.DecimalSI)
-			deltaMemory = resource.NewQuantity(deltaMemory.Value()-oldMemory.Value(), resource.DecimalSI)
-		}
 
-		hardMem, hardMemFound := status.Hard[api.ResourceMemory]
-		if hardMemFound {
-			if set[api.ResourceMemory] && resourcequotacontroller.IsPodMemoryUnbounded(pod) {
-				return false, fmt.Errorf("Limited to %s memory, but pod has no specified memory limit", hardMem.String())
+			// if we do not yet know how much of the current resource is used, we cannot accept any request
+			hard, hardFound := status.Hard[resourceName]
+			if !hardFound {
+				continue
 			}
-			used, usedFound := status.Used[api.ResourceMemory]
+
+			used, usedFound := status.Used[resourceName]
 			if !usedFound {
-				return false, fmt.Errorf("Quota usage stats are not yet known, unable to admit resource until an accurate count is completed.")
+				return false, fmt.Errorf("Unable to admit pod until quota usage stats are calculated.")
 			}
-			if used.Value()+deltaMemory.Value() > hardMem.Value() {
-				return false, fmt.Errorf("Limited to %s memory", hardMem.String())
-			} else {
-				status.Used[api.ResourceMemory] = *resource.NewQuantity(used.Value()+deltaMemory.Value(), resource.DecimalSI)
-				dirty = true
+
+			// the amount of resource being requested, or an error if it does not make a request that is tracked
+			pod := obj.(*api.Pod)
+			delta, err := resourcequotacontroller.PodRequests(pod, resourceName)
+
+			if err != nil {
+				return false, fmt.Errorf("Must make a non-zero request for %s since it is tracked by quota.", resourceName)
 			}
-		}
-		hardCPU, hardCPUFound := status.Hard[api.ResourceCPU]
-		if hardCPUFound {
-			if set[api.ResourceCPU] && resourcequotacontroller.IsPodCPUUnbounded(pod) {
-				return false, fmt.Errorf("Limited to %s CPU, but pod has no specified cpu limit", hardCPU.String())
+
+			// if this operation is an update, we need to find the delta usage from the previous state
+			if a.GetOperation() == admission.Update {
+				oldPod, err := client.Pods(a.GetNamespace()).Get(pod.Name)
+				if err != nil {
+					return false, err
+				}
+
+				// if the previous version of the resource made a resource request, we need to subtract the old request
+				// from the current to get the actual resource request delta
+				oldRequest, err := resourcequotacontroller.PodRequests(oldPod, resourceName)
+				if err != nil {
+					err = delta.Sub(*oldRequest)
+					if err != nil {
+						return false, err
+					}
+				}
 			}
-			used, usedFound := status.Used[api.ResourceCPU]
-			if !usedFound {
-				return false, fmt.Errorf("Quota usage stats are not yet known, unable to admit resource until an accurate count is completed.")
+
+			newUsage := used.Copy()
+			newUsage.Add(*delta)
```

Comment:
I may be missing something, but this code looks correct to me. `delta` is the request if this is a create, and request - old pod's request if this is an update.